### PR TITLE
[RISCV] Select and/or/xor with certain constants to Zbb ANDN/ORN/XNOR

### PR DESCRIPTION
**Pull request by pfusik**
(and X, (C<<12|0xfff)) -> (ANDN X, ~C<<12)
    (or  X, (C<<12|0xfff)) -> (ORN  X, ~C<<12)
    (xor X, (C<<12|0xfff)) -> (XNOR X, ~C<<12)

Emits better code, typically by avoiding an `ADDI HI, -1` instruction.

Co-authored-by: Craig Topper <craig.topper@sifive.com>

### CODE REVIEW

**Comment by pfusik** [comment]
I'd like to implement an optimization so that

```c
long orlow(long x) {
    return x | ((1L << 24) - 1);
}
```

that is now compiled into:

```
        lui     a1, 4096
        addiw   a1, a1, -1
        or      a0, a0, a1
        ret
```

gets instead compiled to:

```
        lui     a1, -4096
        orn     a0, a0, a1
        ret
```

and similarly for AND and XOR.

I'm trying to implement this with tablegen patterns, but I'm new to that technology and I got stuck with the following error:

```
[16/130 -j3] Building RISCVGenGlobalISel.inc...
FAILED: lib/Target/RISCV/RISCVGenGlobalISel.inc /home/p.fusik/upstream/llvm-project/build/lib/Target/RISCV/RISCVGenGlobalISel.inc
cd /home/p.fusik/upstream/llvm-project/build && /home/p.fusik/upstream/llvm-project/build/bin/llvm-tblgen -gen-global-isel -I /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV -I/home/p.fusik/upstream/llvm-project/build/include -I/home/p.fusik/upstream/llvm-project/llvm/include -I /home/p.fusik/upstream/llvm-project/llvm/lib/Target /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVGISel.td --write-if-changed -o lib/Target/RISCV/RISCVGenGlobalISel.inc -d lib/Target/RISCV/RISCVGenGlobalISel.inc.d
anonymous_10632:        (or:{ *:[i32] m1:[i64] } (vt:{} GPR:{ *:[i16 i32] m1:[i16 i32 i64 f64] }:$rs1), (imm:{ *:[i32] m1:[i64] })<<P:Predicate_simm32fff>>:$imm)
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVGISel.td:16:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCV.td:36:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfo.td:2113:
/home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:499:1: error: In anonymous_10632: vt node requires exactly 0 operands!
def : PatGprSimm32fff<or, ORN>;
^
anonymous_10632:        (or:{ *:[i32] m1:[i64] } (vt:{} GPR:{ *:[i16 i32] m1:[i16 i32 i64 f64] }:$rs1), (imm:{ *:[i32] m1:[i64] })<<P:Predicate_simm32fff>>:$imm)
anonymous_10632:        (ORN:{ *:[i32] m1:[i64] } GPR:{ *:[i16 i32] m1:[i16 i32 i64 f64] }:$rs1, (NotImm:{ *:[i16 i32] m1:[i16 i32 i64 f64] } (imm:{ *:[i32] m1:[i64] })<<P:Predicate_simm32fff>>:$imm))
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVGISel.td:16:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCV.td:36:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfo.td:2113:
/home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:499:1: error: In anonymous_10632: Could not infer all types in pattern result!
def : PatGprSimm32fff<or, ORN>;
^
```

Please let me know how can I fix it or if perhaps this approach is completely wrong.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZb.td
Comment (mshockwave) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:494]: `vt` by itself is a SDNode representing ValueType (https://github.com/llvm/llvm-project/blob/1c16807d0dd740ace0f21ed29d1381a0078f745e/llvm/include/llvm/Target/TargetSelectionDAG.td#L367). For example, `sext_inreg` has a vt operand attached at the end to specify the original value type. Therefore you shouldn't put it here and that's what the error message told you.

You might see lots of `vt` popping up here and there in other parts of the backend, and that's because we (for better or for worse) overloading this word for variable names as well. In which case it's just a variable for plugging in different value types.

I think what you want here might be `(XLenVT GPR:$rs1)`?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:494]: `GPR:$rs1` instead of `(XLenVT GPR:$rs1)` should be find since `XLenVT` is already on the outside of `OpNode`.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZb.td
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:495]: We need to explicitly create an LUI for the immediate here. And we need to put the immediate into LUI format which requires shifting right by 12 and masking to 20 bits.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:495]: MachineLICM pass should be able to move it outside loops. SelectionDAG will CSE it if its used multiple times in the same basic block. Otherwise MachineCSE pass will be responsible for CSE.

**Comment by topperc** [comment]
Here's another case that might be useful to optimize https://godbolt.org/z/MEzP15sas it already generates a `not` but we don't fold it.

**Comment by pfusik** [comment]
Force pushed because of a precommit test.

File: llvm/test/CodeGen/RISCV/pr84653_pr85190.ll
Comment (pfusik) [inline llvm/test/CodeGen/RISCV/pr84653_pr85190.ll:23]: `lui` is unchanged here because it's a `X ^ ((1 << 31) - 1) -> X ^ ~(1 << 31)` case.

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (pfusik) [inline llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:67]: In this case we pessimize. I expect this to be rare. Do we want to address it? How?

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (pfusik) [inline llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:93]: Ditto.

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (pfusik) [inline llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:242]: With Zbs this is:

        ori     a0, a0, 2047
        bseti   a0, a0, 11

and is not affected by this change. Shall I add `+zbs` or test both with and without?

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (pfusik) [inline llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:252]: Similar to the above.

        xori    a0, a0, 2047
        binvi   a0, a0, 11

**Comment by pfusik** [comment]
> Here's a rough sketch using a ComplexPattern instead of PatLeaf/SDNodeXForm.

This works great!

I didn't even have to special-case `and` for `sll+srl` etc. But I don't know why. Does the pattern order matter or is there some scoring mechanism?

**Comment by pfusik** [comment]
> Here's another case that might be useful to optimize https://godbolt.org/z/MEzP15sas it already generates a `not` but we don't fold it.

I can work on it.

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (lenary) [reply llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:67]: Can you do a `hasOneUse` check on the ConstantSDNode? I don't know whether it's worth trying.

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (lenary) [reply llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:242]: I would add RUN lines with `+zbs`, rather than adding `+zbs` to the existing RUN lines.

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (pfusik) [reply llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:67]: Yes, this prevents pessimizations in `oraddlow16` and `addorlow16`. But also disables optimization in `orarray3`.

Comment (lenary) [reply llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll]: I did think about this, but I'm not sure whether we want to be looking at all the users like that during selection, or whether this should be a target-specific DAGCombine rather than a set of selection patterns. I'll defer to craig on that.

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:3244]: Push the `!` into the individual conditions and use `||`.

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll]: I think I prefer checking the users during selection. If we do a target specific combine we need new ISD nodes to represent the optimization.

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (pfusik) [inline llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:299]: Here it's possible to optimize a non-int32. Getting the condition right in general seems tricky. Can I call `RISCVMatInt` for `Imm` and `~Imm` to see which one is shorter?

**Review by pfusik — COMMENTED** [review]
Added Zbs testing. Added `andimm64` test. Squashed the change so that the effect on tests is clear.

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp
Comment (pfusik) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp]: Done

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (pfusik) [reply llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:242]: Done

File: llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll
Comment (lenary) [reply llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:299]: This wouldn't be the only place to compare two `getIntMatCost`s, so it should be fine, as long as you check the simpler exclusionary conditions first - maybe even including your `isBitwiseLogicOp` loop.

**Comment by pfusik** [comment]
> Here's another case that might be useful to optimize https://godbolt.org/z/MEzP15sas it already generates a `not` but we don't fold it.

Done. Covered by the `andimm64srli` test.

**Review by lenary — APPROVED** [review]
I'm happy with this.

By default, LLVM will squash-merge, so can you pre-commit the new test file before landing this?

**Comment by pfusik** [comment]
> By default, LLVM will squash-merge, so can you pre-commit the new test file before landing this?

Do you mean create a new PR for the pre-commit test?

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:3247]: This needs to change `N->users()` now.

**Comment by topperc** [comment]
> > By default, LLVM will squash-merge, so can you pre-commit the new test file before landing this?
> 
> Do you mean create a new PR for the pre-commit test?

If you have commit access you can directly push it without going through github.

**Comment by lenary** [comment]
Nice work!

**Comment by pfusik** [comment]
Thank you.

**Pull request merged**
